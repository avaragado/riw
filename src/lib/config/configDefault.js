// @flow

const cfg: RIWConfig = {
    configFile: null, // overridden
    rootDir: '', // overridden
    defaultLocale: 'en-us',
    targetLocales: [],
    translationsDatabaseFile: 'src/locale/riw-db.json',
    sourceDirs: ['src/**/*.js'],
    collateDir: 'tmp/babel-plugin-react-intl',
    inputMode: 'source',
    translationsOutputFile: 'src/locale/[locale].json',
    outputMode: 'file-per-locale',
    todoFile: 'src/locale/TODO-untranslated.json',
};

export default cfg;
