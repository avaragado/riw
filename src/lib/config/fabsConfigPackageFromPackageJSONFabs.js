// @flow

import path from 'path';

export default (fabsPackageJSON: ?AbsolutePath): ?AbsolutePath => (
    fabsPackageJSON
        ? path.resolve(path.dirname(fabsPackageJSON), '.riw-config.js')
        : null
    );
