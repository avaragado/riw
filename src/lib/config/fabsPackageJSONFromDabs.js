// @flow

import findUp from 'find-up';

export default (cwd: AbsolutePath): ?AbsolutePath => findUp.sync('package.json', { cwd });

