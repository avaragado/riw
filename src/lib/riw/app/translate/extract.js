// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import prop from 'ramda/src/prop';

import { arfabsInputJSON, arfabsInputSource } from '../../../config-helper';

import armdFromJSONFabs from './armdFromJSONFabs';
import armdfromSourceFabs from './armdFromSourceFabs';

type MDExtractor = (arfabsFromConfig: FilesFromConfig, armdFromFabs: MessageDescriptorsFromFile) =>
    <T>(notify: string => T => T) =>
    (config: RIWConfig) => RIWMessageDescriptor[];

const armdExtract: MDExtractor = (arfabsFromConfig, armdFromFabs) => notify => compose(
    notify('endExtract'),
    chain(prop('armd')),
    map(compose(
        notify('endExtractFile'),
        fabs => ({
            fabs,
            armd: map(
                md => ({ ...md, file: fabs }),
                armdFromFabs(fabs),
            ),
        }),
        notify('startExtractFile'),
    )),
    notify('startExtract'),
    arfabsFromConfig,
);

export const armdExtractSource = armdExtract(arfabsInputSource, armdfromSourceFabs);
export const armdExtractJSON = armdExtract(arfabsInputJSON, armdFromJSONFabs);
