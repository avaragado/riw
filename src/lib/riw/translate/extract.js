// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import prop from 'ramda/src/prop';

import * as source from './extractFromSource';
import * as json from './extractFromJSON';

type MDExtractor = (arfabsFromConfig: FilesFromConfig, armdFromFabs: MessageDescriptorsFromFile) =>
    <T>(notify: string => T => T) =>
    (config: RIWConfig) => RIWMessageDescriptor[];

const armdExtract: MDExtractor = (arfabsFromConfig, armdFromFabs) => notify => compose(
    notify('endExtract'),
    chain(prop('armd')),
    map(compose(
        notify('endExtractFile'),
        fabs => ({
            fabs,
            armd: map(
                md => ({ ...md, fabs }),
                armdFromFabs(fabs),
            ),
        }),
        notify('startExtractFile'),
    )),
    notify('startExtract'),
    arfabsFromConfig,
);

export const armdExtractSource = armdExtract(source.arfabsFromConfig, source.armdFromFabs);
export const armdExtractJSON = armdExtract(json.arfabsFromConfig, json.armdFromFabs);
