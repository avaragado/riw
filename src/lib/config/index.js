// @flow

import configDefault from './resolve/configDefault';
import fabsConfigPackageForCWD from './fromPackage/fabsConfigPackageForCWD';
import configFromPath from './fromPath';
import configFromPackage from './fromPackage';
import configResolve from './resolve';

export {
    configDefault,
    fabsConfigPackageForCWD,
    configFromPath,
    configFromPackage,
    configResolve,
};
