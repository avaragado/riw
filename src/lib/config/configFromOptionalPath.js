// @flow

import path from 'path';

import log from '../log';

import configDefault from './configDefault';
import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

const requireOrNull = (fabs: AbsolutePath): ?Object => {
    try {
        return require(fabs);

    } catch (err) {
        return null;
    }
};

const configFromPath = (pathConfig: Path): ?RIWConfig => {
    const fabs: AbsolutePath = path.resolve(pathConfig);
    const configOverride = requireOrNull(fabs);

    if (configOverride === null) {
        log.error('riw', 'File not found: %j.', fabs);
        return null;
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabs),
    };
};

const configFromPackage = (): ?RIWConfig => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsConfigPackage || !fabsPackageJSON) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    let configOverride: ?Object = requireOrNull(fabsConfigPackage);

    if (configOverride === null) {
        configOverride = requireOrNull(fabsPackageJSON);

        if (configOverride && configOverride.riw) {
            configOverride = configOverride.riw;
        } else {
            configOverride = null;
        }
    }

    if (configOverride === null) {
        log.warn('riw', 'No config located for the project. You probably want to add some configuration.');
        configOverride = {};
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabsConfigPackage),
    };

};

export default (pathConfig: ?Path): ?RIWConfig => (
    pathConfig
        ? configFromPath(pathConfig)
        : configFromPackage()
);
