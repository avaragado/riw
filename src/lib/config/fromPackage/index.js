// @flow

import log from '../../log';

import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

export default (): ?RIWConfigSparseWithSource => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsPackageJSON || !fabsConfigPackage) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    try {
        return {
            ...(require(fabsConfigPackage): RIWConfigSparse),
            configFile: fabsConfigPackage,
        };

    } catch (err) {
        // fabsPackageJSON exists as we hunted the fs for it above:
        // but it might not have the riw key.
        const config: RIWConfigSparse = require(fabsPackageJSON).riw;

        if (!config) {
            log.warn('riw', 'No app config found. You probably want to add some configuration.');
        }

        return {
            ...config,
            configFile: fabsPackageJSON,
        };
    }
};
