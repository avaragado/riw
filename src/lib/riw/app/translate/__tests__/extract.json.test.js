// @flow

import path from 'path';

import mock from 'mock-fs';
import outdent from 'outdent';

import type { ConfigSparse, ConfigSparseWithSource } from '../../../../config';
import { configResolve } from '../../../../config';

import { armdExtractJSON } from '../extract';

const cfgBase: ConfigSparseWithSource = {
    rootDir: '/fixtures',
    sourceDirs: ['dir/**/*.js'],
};

const notify = () => x => x;

type Fixture = {
    name: string,
    in: { [key: string]: string },
    configOverride?: ConfigSparse,
};

const fixtures: Fixture[] = [
    {
        name: '01',
        in: {}, // empty dir
    },

    {
        name: '02',
        in: {
            'a.json': '[]',
            'b.json': '[]',
        },
    },

    {
        name: '03',
        in: {
            'a.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1!"
                    }
                ]
            `,
            'aa.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1 again!"
                    }
                ]
            `,
            'b.json': outdent`
                [
                    {
                        "id": "b.1",
                        "defaultMessage": "b.1!"
                    },
                    {
                        "id": "b.2",
                        "description": "b.2 desc",
                        "defaultMessage": "b.2!"
                    }
                ]
            `,
            'bb.json': outdent`
                [
                    {
                        "id": "b.2",
                        "description": "b.2 again desc",
                        "defaultMessage": "b.2 again!"
                    }
                ]
            `,
            'c.json': outdent`
                [
                    {
                        "id": "c.1",
                        "defaultMessage": "c.1!"
                    }
                ]
            `,
        },
        configOverride: {
            collateDir: 'dir',
            inputMode: 'json',
        },
    },

    {
        name: '04',
        in: {
            'a.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1!"
                    }
                ]
            `,
            'aa.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1 again!"
                    }
                ]
            `,
            'b.json': outdent`
                [
                    {
                        "id": "b.1",
                        "defaultMessage": "b.1!"
                    },
                    {
                        "id": "b.2",
                        "description": "b.2 desc",
                        "defaultMessage": "b.2!"
                    }
                ]
            `,
            'bb.json': outdent`
                [
                    {
                        "id": "b.2",
                        "description": "b.2 again desc",
                        "defaultMessage": "b.2 again!"
                    }
                ]
            `,
            'c.json': outdent`
                [
                    {
                        "id": "c.1",
                        "defaultMessage": "c.1!"
                    }
                ]
            `,
        },
        configOverride: {
            collateDir: path.resolve('/fixtures', 'dir'),
            inputMode: 'json',
        },
    },
];

describe('lib/riw/app/translate/extract.json', () => {
    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                '/fixtures': {
                    dir: fixture.in,
                },
            });

            const cfg = configResolve({
                ...cfgBase,
                ...fixture.configOverride,
            });

            const received = armdExtractJSON(notify)(cfg);

            mock.restore();

            expect(received).toMatchSnapshot();
        });
    });
});
