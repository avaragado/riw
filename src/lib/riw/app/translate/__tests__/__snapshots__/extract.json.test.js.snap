// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/app/translate/extract.json 01 1`] = `Array []`;

exports[`lib/riw/app/translate/extract.json 02 1`] = `Array []`;

exports[`lib/riw/app/translate/extract.json 03 1`] = `
Array [
  Object {
    "defaultMessage": "a.1!",
    "file": "/fixtures/dir/a.json",
    "id": "a.1",
  },
  Object {
    "defaultMessage": "a.1 again!",
    "file": "/fixtures/dir/aa.json",
    "id": "a.1",
  },
  Object {
    "defaultMessage": "b.1!",
    "file": "/fixtures/dir/b.json",
    "id": "b.1",
  },
  Object {
    "defaultMessage": "b.2!",
    "description": "b.2 desc",
    "file": "/fixtures/dir/b.json",
    "id": "b.2",
  },
  Object {
    "defaultMessage": "b.3!",
    "description": Object {
      "abc": "def",
      "foo": "bar",
    },
    "file": "/fixtures/dir/b.json",
    "id": "b.3",
  },
  Object {
    "defaultMessage": "b.2 again!",
    "description": "b.2 again desc",
    "file": "/fixtures/dir/bb.json",
    "id": "b.2",
  },
  Object {
    "defaultMessage": "c.1!",
    "file": "/fixtures/dir/c.json",
    "id": "c.1",
  },
]
`;

exports[`lib/riw/app/translate/extract.json 04 1`] = `
Array [
  Object {
    "defaultMessage": "a.1!",
    "file": "/fixtures/dir/a.json",
    "id": "a.1",
  },
  Object {
    "defaultMessage": "a.1 again!",
    "file": "/fixtures/dir/aa.json",
    "id": "a.1",
  },
  Object {
    "defaultMessage": "b.1!",
    "file": "/fixtures/dir/b.json",
    "id": "b.1",
  },
  Object {
    "defaultMessage": "b.2!",
    "description": "b.2 desc",
    "file": "/fixtures/dir/b.json",
    "id": "b.2",
  },
  Object {
    "defaultMessage": "b.2 again!",
    "description": "b.2 again desc",
    "file": "/fixtures/dir/bb.json",
    "id": "b.2",
  },
  Object {
    "defaultMessage": "c.1!",
    "file": "/fixtures/dir/c.json",
    "id": "c.1",
  },
]
`;
